Allow using existing Keep-A-Changelog files for release notes

weierophinney opened the pull request:
This patch modifies the workings of the `ReleaseCommand` slightly.

It now accepts a new argument `Laminas\AutomaticReleases\Changelog\ReleaseChangelog $releaseChangelog`. The `ReleaseChangelog` interface is intended to prepare a changelog in Keep-A-Changelog format for release (generally by adding the release date), commit it, and push it back to the origin repository.

To accomplish this, I created:

- `CommitFile`, which is intended for, you guessed it, committing a file to the local checkout. One implementation was provided, `CommitFileViaConsole`.
- `Laminas\AutomaticReleases\Changelog\ReleaseChangelogViaKeepAChangelog`, which uses tooling from phly/keep-a-changelog to add the release date to the `CHANGELOG.md` file, and then a `CommitFile` and `Push` instance to commit and push the changes. If the file does not exist, is in a different format, or has no changes necessary, it acts as a no-op.

To allow more ways to generate the release text, I added two additional arguments to the `CreateReleaseText` interface: `BranchName $sourceBranch` and `string $repositoryPath`. Additionally, I added a `canCreateReleaseText(): bool` method that accepts the same arguments as `__invoke()`; this allows testing to see if a given implementation will be able to create the release text.

From there, I added:

- `CreateReleaseTextViaKeepAChangelog`. It's `canCreateReleaseText()` method returns `false` if the file does not exist, is malformed, or does not have an entry corresponding to the version. Otherwise, the `__invoke()` method extracts the changelog entry using the phly/keep-a-changelog `ChangelogParser`.
- `AppendingCreateReleaseTextAggregate`, which accepts a list of `CreateReleaseText` instances. It filters for those capable of creating release text, and then concatenates them with `\n\n-----\n\n`.

When creating the `ReleaseCommand` in `bin/console.php`, I now create an `AppendingCreateReleaseTextAggregate` that uses a `CreateReleaseTextViaKeepAChangelog` instance and a `CreateReleaseTextThroughChangelog` instance.

Fixes #5

weierophinney committed:
qa: add phly/keep-a-changelog as a dependency

To be used as an alternate branch when a CHANGELOG file is available.

Requires switching from ocramius/package-versions to
composer/package-versions-deprecated.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
WIP: add ability to use Keep-A-Changelog to set release date in CHANGELOG and fetch CHANGELOG contents

This patch modifies the workings of the `ReleaseCommand` slightly.

Instead of depending on the `CreateReleaseText` interface, it now
depends on a new `Laminas\AutomaticReleases\Changelog\ReleaseChangelogAndFetchContents`
interface. This new interface consumes a `ReleaseChangelogEvent`, which
stores a variety of runtime-determined values (console input/output,
milestone, repository name and path, etc.). Implementors of this
interface pull values from the event that they might use to fetch or
create the changelog and return it. If they are incapable of doing so,
they return null.

I have created three implementations:

- `UseKeepAChangelogEventsToReleaseAndFetchChangelog`, which checks to
  see if there is a `CHANGELOG.md` file in the repository. If not, it
  returns null immediately. Otherwise, it sets the release date in the
  changelog for the given version, commits the file (more on that
  later), and pushes the changes to the release branch. It then returns
  the discovered changelog contents

- `CreateChangelogViaMilestone` pulls data from the event for use with
  the original `CreateReleaseText` implementation, and decorates such an
  instance.

- `ReleaseChangelogAndFetchContentsAggregate` takes a list of
  `ReleaseChangelogAndFetchContents` instances, and returns the first
  changelog discovered. If none return one, it generates a basic one using
  just the repository name and release version.

I have updated `bin/console.php` to create a
`ReleaseChangelogAndFetchContentsAggregate` instance composing each of
the other two types, with the keep-a-changelog version being first.

To assist in how the `UseKeepAChangelogEventsToReleaseAndFetchChangelog`
implementation works, I created `Laminas\AutomaticReleases\Git\CommitFile`,
which describes a commit operation on a single file. `Laminas\AutomaticReleases\Git\CommitFileViaConsole`
does a `git add` operation followed by a `git commit` operation on the
file.

Currently, no tests; I needed to play with architecture to figure out a
workable approach first.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: adapted ReleaseCommandTest to mirror changes to ReleaseCommand

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: apply phpcs fixes

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on composer.json:
```diff
@@ -10,9 +10,10 @@
     },
     "require": {
         "php": "^7.4.7",
+        "composer/package-versions-deprecated": "^1.10.99",
```
Note: we're already on `^7.4.7`, so a dependency change shouldn't be needed here

weierophinney replied:
It's needed because phly/keep-a-changelog uses it; otherwise, there are conflicts.

boesing replied:
Yeah, something messed up with that replacement. Had some weird problems when using laminas-cli aswell...
Not sure what causes those conflicts...

Ocramius replied:
Could be any of the CLI tools - not a big deal though, as we're free to use `composer/composer:^2` here (it's an application, not a library)

Ocramius commented on src/Git/CommitFile.php:
```diff
@@ -0,0 +1,15 @@
+        string $repositoryDirectory,
+        string $filename,
+        string $commitMessage,
+        string $author
```
This shouldn't be needed, in theory, given that the repository is pre-configured with an author email and name?

Ocramius replied:
Note also that git will honor any environment variables regarding current author/email

weierophinney replied:
I was unclear if this were the case, but seeing how the `CreateTagViaConsole` class is written, it must be. I'll revert that in my next revisions.

Ocramius commented on src/Changelog/ReleaseChangelogAndFetchContents.php:
```diff
@@ -0,0 +1,13 @@
+
+namespace Laminas\AutomaticReleases\Changelog;
+
+interface ReleaseChangelogAndFetchContents
```
The purpose of this interface seems to be unclear and/or overloaded.

I think we should instead:

 1. generate the changelog text
 2. replace it in the pre-existing `CHANGELOG.md` if applicable

Those should be two steps, where one is a pure read operation, and the other is a read+write operation

Ocramius commented on src/Changelog/ReleaseChangelogEvent.php:
```diff
@@ -0,0 +1,44 @@
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+final class ReleaseChangelogEvent
```
This should be marked with `/** @psalm-immutable */`

In addition to that, `InputInterface` and `OutputInterface` should not be part of it, as they are part of the application layer of this system, and mutable objects too.

Overall, this object seems to contain too much state that obfuscates the real dependency of consumers of it: splitting it into parameters will probably highlight some areas with too much complexity in them.

Ocramius commented on src/Changelog/UseKeepAChangelogEventsToReleaseAndFetchChangelog.php:
```diff
@@ -0,0 +1,73 @@
+use function file_exists;
+use function sprintf;
+
+class UseKeepAChangelogEventsToReleaseAndFetchChangelog implements
```
The `CreateChangelogViaMilestone` and `UseKeepAChangelogEventsToReleaseAndFetchChangelog` don't really do the same things:

 * one creates a commit on the repo (this class)
 * the other (`UseKeepAChangelogEventsToReleaseAndFetchChangelog`) produces text to be used for the release

They should not share the same interface, and should not be put together under `ReleaseChangelogAndFetchContentsAggregate`.

In fact, the nullability in `    public function __invoke(ReleaseChangelogEvent $releaseChangelogEvent): ?string` indicates profound differences in behavior, and force you to implement a fallback mechanism if generating a changelog failed.

I think we should:

 1. use keep a changelog to generate `CHANGELOG.md` (and we do use it explicitly before creating a tag)
 2. use `CreateChangelogViaMilestone` to generate the tag description and release notes (same as it was before, relatively UI friendly for github, and still references a commit that contains a `CHANGELOG.md` diff)

Alternatively, we can either:

 * append `CreateChangelogViaMilestone` to tag description/release name, in addition to what `UseKeepAChangelogEventsToReleaseAndFetchChangelog` generated
 * scrap `CreateChangelogViaMilestone`, and generate everything with `UseKeepAChangelogEventsToReleaseAndFetchChangelog`

This means reverting some of these abstractions, and calling the logic sequentially:

```php
$keepAChangelogText = ($this->keepAChangelog)(...);
$releaseText = ($this->createChangelogViaMilestone)(...);

$tag(..., $releaseText . "\n\n" . $keepAChangelogText);
```

No need to aggregate these disjunct pieces of logic together, and no need to change pre-existing tests (besides CLI command tests)

weierophinney replied:
I disagree with some of this assessment.

KeepAChangelog does not **generate** a changelog; it **is** the changelog. The phly/keep-a-changelog tooling was being used to:

- Set the release date for the given changelog version.
- Retrieve the contents for that changelog version for purposes of the GitHub release.

That first part is where we have an overlap of responsibilities, as it also involves creating a commit.

The reason I want the Keep-A-Changelog format in the GitHub release is because if you search for `{package} {version} changelog`, you end up at the GitHub releases page, which is where you'll then read what _changed_ for a given version. The human-readable format is important here.

That said, I really like what JwageGenerateChangelog creates, as it provides _attribution_, which is also something I think it's important for us to do as a community OSS project.

On top of all that: I can see value in this tooling outside of Laminas. Yes, we're the primary consumers right now, but you've done a great job of engineering it such that it is generalized enough to work with any project that wants to adopt the workflow it implements for tagging, releasing, and up-merging. As such, we cannot necessarily assume that a given project has a changelog, much less in keep-a-changelog format.

So, here's my idea for moving forward:

- I'll split out the "add date to changelog and commit file" aspect into a separate utility that is passed to `ReleaseCommand` and invoked prior to retrieving the changelog text. This utility will be a no-op if no changelog file is present or if the changelog is not in keep-a-changelog format.

- I'll merge the `ReleaseChangelogAndFetchContents` into `CreateReleaseText` interfaces, adding arguments for source branch and repository directory to the latter, and make the current `UseKeepAChangelog....` implement that instead, having it return an empty string if no keep-a-changelog file is found. I'll also create an "appending aggregate" version that executes each and concatenates it to the previous.

I think this will resolve the architectural issues you've flagged.

Ocramius replied:
> KeepAChangelog does not generate a changelog; it is the changelog. The phly/keep-a-changelog tooling was being used to:

One is `f :: API -> Text`, the other is `f :: API -> (Text, GitCommit)` - very different underlying architecture

Ocramius replied:
> I think this will resolve the architectural issues you've flagged.

Will likely be able to do another review tonight :)

Ocramius replied:
Discussed in chat. It is clear now that this should be a 2 step operation of: 

1. generating the text (to be manipulated and/or used directly)
2. committing it to repo (separate service/function that does just that)

weierophinney committed:
qa: suppress psalm deprecation notice in console file

While composer/package-versions-deprecated is deprecated with Composer
2, we will be using it for the forseeable future. As such, it's an
expected issue, and should be suppressed.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: ensure proper types are used everywhere

- Updates to phly/keep-a-changelog 2.8, which adds typed accessors to its
  `ChangelogEntry` class, which is used internally here, and for which
  psalm was raising errors due to inability to determine types.

- Updates `bin/console.php` to create a `$createCommitText` variable
  assigned to a `CreateReleaseTextThroughChangelog` instance. This
  instance is then composed in the `$createReleaseText` array, and used
  directly for the `CreateMergeUpPullRequest`, where that changelog type
  makes sense as part of a commit message.

- Adds `@psalm-*` annotations to better detail properties, arguments,
  and variables where necessary.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: unit tests for CreateChangelogViaMilestone class

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: unit tests for ReleaseChangelogAndFetchContentsAggregate

Demonstrates the behaviors of
`ReleaseChangelogAndFetchContentsAggregate`. Also modifies the return
type to only `string` (instead of being nullable), as the aggregate will
always return a string, which follows PHP covariance rules.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: adds unit tests for CommitFileViaConsole

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: adds unit tests for UseKeepAChangelogEventsToReleaseAndFetchChangelog

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: apply CS fixes and psalm advice

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: remove author awareness from CommitFile and ReleaseChangelogEvent

As noted during review, git honors any environment variables defining
the current author and email, so there's no need to pass them in.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: remove input/output awareness from ReleaseChangelogEvent

Removes input/output awareness from the event; we can use dummy
instances when triggering the event later, as the input and output are
not consulted.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: accept a branch name when committing a file

Ensures that we commit against the correct branch.
`CommitFileViaConsole` is now updated to checkout the given branch
before adding and committing the file.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: new interface, ReleaseChangelog

Describes classes that can prepare a changelog file for release.

One implementation: ReleaseChangelogViaKeepAChangelog, which uses
phly/keep-a-changelog tools to determine (a) if there is a changelog to
release, and then (b) updates the date in it, commits the file, and
pushes it.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: idempotent creation of release text via keep-a-changelog

Updates `CreateReleaseText` to:

- Add a `BranchName $sourceBranch` argument
- Add a `string $repositoryDir` argument
- Add a new `canCreateReleaseText()` method with identical arguments,
  returning a boolean. The method should be called before invoking the
  implementation to create release text.

After making those changes, updated `CreateReleaseTextThroughChangelog`
to add the arguments and new method, and added a test for the new
method.

Finally, adds a new class, `CreateReleaseTextThroughChangelog`, which
uses the phly/keep-a-changelog `ChangelogParser` to determine if (a) it
can locate a CHANGELOG.md file and changelog entry matching the release
version, and then (b) pull it and return it.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: creates AppendingCreateReleaseTextAggregate as aggregate CreateReleaseText implementation

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: add ReleaseChangelog as ReleaseCommand argument and switch back to CreateReleaseText

Adds a new dependency to the `ReleaseCommand`, `ReleaseChangelog
$releaseChangelog`, and reverts the `$createChangelogText` argument to
the `CreateReleaseText` type. Internally, it now executes the
`$releaseChangelog` prior to calling `$createReleaseText`.

In `bin/console.php`, we now create an
`AppendingCreateReleaseTextAggregate` to inject as the
`$createReleaseText` argument of `ReleaseCommand`, using each of
`CreateReleaseTextViaKeepAChangelog` and
`CreateReleaseTextThroughChangelog` instances, in that order.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
It looks like the failing unit test is likely due to a missing git author/email, but it's hard to determine. Tests pass fine locally, which is why I'm inferring it's an environment issue.

weierophinney commented:
> It looks like the failing unit test is likely due to a missing git author/email, but it's hard to determine.

Verified: https://github.com/laminas/automatic-releases/pull/18/checks?check_run_id=946634877#step:8:68

Do you want me to add the git author/email to the travis env as part of this patch?

weierophinney commented:
@Ocramius Okay, ready for another round of review. :smile:

weierophinney commented on test/unit/Changelog/ReleaseChangelogViaKeepAChangelogTest.php:
```diff
@@ -0,0 +1,212 @@
+        return $repo;
+    }
+
+    private const INVALID_CHANGELOG = <<< 'END'
+        # NOT A CHANGELOG
+
+        This file is not a changelog.
+
+        ## Bad headers
+
+        It contains bad headers, among other things.
+
+        END;
```
Since we're on 7.4... One cool feature of 7.3 is that it will now strip indentation from heredoc/nowdoc contents if the end marker is indented. Makes these constant declarations far more readable!

Ocramius replied:
I would still endorse using a different start/end marker depending on context. For example, here I would have used `<<< 'MARKDOWN'`

Not important, since we're in the middle of a relatively clear test

weierophinney commented on test/unit/Github/CreateReleaseTextViaKeepAChangelogTest.php:
```diff
@@ -0,0 +1,204 @@
+            ### Fixed
+            
+            - Nothing.
+            END, $date);
```
Also, heredoc/nowdoc no longer require the end marker to be on a line of their own followed by a `;`, which makes this work.

Ocramius replied:
Oooh, didn't know! Awesome, but it does indeed confuse the heck out of the github PHP parsing :D

weierophinney commented on src/Github/AppendingCreateReleaseTextAggregate.php:
```diff
@@ -0,0 +1,77 @@
+        BranchName $sourceBranch,
+        string $repositoryDirectory
+    ): string {
+        $items = array_map(
+            static fn (CreateReleaseText $strategy): string => $strategy($milestone, $repositoryName, $semVerVersion, $sourceBranch, $repositoryDirectory),
+            array_filter(
+                $this->strategies,
+                // phpcs:disable
+                fn(CreateReleaseText $strategy): bool => $strategy->canCreateReleaseText($milestone, $repositoryName, $semVerVersion, $sourceBranch, $repositoryDirectory)
+                // phpcs:enable
+            )
```
I used arrow functions here so I wouldn't need to use `use(...)` with that long list of arguments.

Ocramius replied:
Not sure why phpcs complains above though :thinking: 

Perhaps it wants `static fn`?

weierophinney commented on src/Changelog/ReleaseChangelogViaKeepAChangelog.php:
```diff
@@ -0,0 +1,141 @@
+         *
+         * @psalm-suppress PropertyNotSetInConstructor
+         */
+        $event = new class ($versionString) extends ReadyLatestChangelogEvent {
```
I used anonymous classes in this implementation as they are one-offs, and only set the values necessary for the specific event listeners I'm manually triggering. If you want me to extract these, let me know.

Ocramius replied:
I think this is an OK trade-off: we can't go and change the whole upstream library for this :)

weierophinney reviewed (commented):
A few comments on implementation details...

Ocramius commented on bin/console.php:
```diff
@@ -56,17 +60,24 @@ static function (int $errorCode, string $message = '', string $file = '', int $l
         $httpClient
     ));
-    $push                 = new PushViaConsole();
+    $createReleaseText    = new AppendingCreateReleaseTextAggregate([
```
The name `AppendingCreateReleaseTextAggregate` is quite terrible - can we synthesise into a small sentence?

weierophinney replied:
How about `AppendSeveralReleaseTexts`?

Ocramius replied:
s/append/concatenate, but otherwise  :+1:

weierophinney replied:
Never mind - I like your suggestion later of `ConcatenateMultipleReleaseTexts`.

Ocramius commented on src/Changelog/ReleaseChangelog.php:
```diff
@@ -0,0 +1,20 @@
+use Laminas\AutomaticReleases\Git\Value\BranchName;
+use Laminas\AutomaticReleases\Git\Value\SemVerVersion;
+
+interface ReleaseChangelog
```
This name is not clear: is it  supposed to be "commit release changelog" perhaps? Since its signature is `: void`, I suppose it **does** something

weierophinney replied:
Yeah, `CommitReleaseChangelog` makes sense - I was thinking "release" as a verb here, but we also have a concept of "a release" throughout. If that name makes sense, I'll update.

Ocramius commented on src/Changelog/ReleaseChangelogViaKeepAChangelog.php:
```diff
@@ -0,0 +1,141 @@
+use Phly\KeepAChangelog\Version\SetDateForChangelogReleaseListener;
+use Symfony\Component\Console\Output\NullOutput;
+
+use function date;
```
This should not be: if the current date/time is needed, let's inject it via a `Clock` (https://github.com/lcobucci/clock is a simple abstraction for it)

weierophinney replied:
As in, inject it as an argument? as a property injected or created via constructor? Creation of the instance would be something like `new FrozenClock(new DateTimeImmutable('now'))` in the `bin/console.php` script?

Essentially, I just need the date string, so internally, where I have `date('Y-m-d')`, this would be come `$clock->now()->format('Y-m-d')`?

Ocramius replied:
in here, `$this->clock->now()->format('Y-m-d')`

In the constructor call, `new SystemClock()` is given as a dependency

Ocramius commented on src/Changelog/ReleaseChangelogViaKeepAChangelog.php:
```diff
@@ -0,0 +1,141 @@
+            $repositoryDirectory,
+            $sourceBranch,
+            'CHANGELOG.md',
+            sprintf('%s readiness', $versionString)
```
`1.2.3 readiness` doesn't sound like an awesome commit text: can we use something that explains that we did just provide a changelog text?

weierophinney replied:
Will do!

Ocramius commented on src/Changelog/ReleaseChangelogViaKeepAChangelog.php:
```diff
@@ -0,0 +1,141 @@
+    {
+        $event = $this->createReadyLatestChangelogEvent($changelogFile, $versionString);
+
+        (new DiscoverChangelogEntryListener())($event);
```
These listeners are bizarre: I know they are mostly imported, and you mostly did C&P of some upstream code: still feels weird for them to mutate `$event`.

Probably design similar to what we had in `laminas/laminas-mvc`?

weierophinney replied:
PSR-14, actually, via the `StoppableEventInterface`. Essentially, dispatchers check for `$event->isPropagationStopped()` after triggering each listener; if it returns true, it doesn't iterate to the next listener. I use this heavily in keep-a-changelog, as it makes it easier for me to report error conditions back to the user and/or to stop doing work if nothing else needs to be done. I use a `failed()` method to check for it in my library, and `isPropagationStopped()` typically checks for that (sometimes it checks other things as well).

Ocramius commented on src/Changelog/ReleaseChangelogViaKeepAChangelog.php:
```diff
@@ -0,0 +1,141 @@
+        SemVerVersion $version,
+        BranchName $sourceBranch
+    ): void {
+        $changelogFile = $repositoryDirectory . '/CHANGELOG.md';
```
`CHANGELOG.md` is referenced twice in here: in order to prevent typos in future, let's move it to a `private const`

weierophinney replied:
Will do.

Ocramius commented on src/Github/AppendingCreateReleaseTextAggregate.php:
```diff
@@ -0,0 +1,77 @@
+use function array_map;
+use function implode;
+
+final class AppendingCreateReleaseTextAggregate implements CreateReleaseText
```
Mentioned above: the naming is a bit weird, but that's because it follows the pattern rather than the actual intent.

Something like `ConcatenateMultipleReleaseTexts` would suffice

Ocramius commented on src/Github/AppendingCreateReleaseTextAggregate.php:
```diff
@@ -0,0 +1,77 @@
+    /** @var CreateReleaseText[] */
+    private array $strategies;
+
+    /** @param CreateReleaseText[] $strategies */
```
`/** @psalm-param non-empty-list<CreateReleaseText> */` and the assertion goes away :D

Ocramius commented on src/Github/AppendingCreateReleaseTextAggregate.php:
```diff
@@ -0,0 +1,77 @@
+    public function __construct(array $strategies)
+    {
+        Assert::notEmpty($strategies);
+        $this->strategies = $strategies;
```
More than `strategies` (pattern name), possibly `$this->releaseTextGenerators` or such?

Ocramius commented on src/Github/AppendingCreateReleaseTextAggregate.php:
```diff
@@ -0,0 +1,77 @@
+            )
+        );
+
+        /** @psalm-var non-empty-string $changelog */
```
Considering that (in theory) no generator may support this, a non-empty-string is not guaranteed.

In order to prevent that, let's add `Assert::notEmpty()` **after** `$changelog` has been assigned.

The inline docblock is coding a too heavy assumption here, so we turn it into an explicit crash in case of misconfiguration.

Ocramius commented on src/Github/CreateReleaseTextThroughChangelog.php:
```diff
@@ -56,6 +59,16 @@ public function __invoke(
+        RepositoryName $repositoryName,
+        SemVerVersion $semVerVersion,
+        BranchName $sourceBranch,
+        string $repositoryDirectory
```
Not important in this patch, so feel free to dismiss (or creating a new issue for me to solve later), but I think we're repeating `$repositoryDirectory` **a lot** across this codebase.

A value type would prevent us from passing a random `non-empty-string` in here.

weierophinney replied:
I'd be willing to add that as part of this patch if you'd like; I rely on it **heavily**, as I need to work with the filesystem.

Ocramius replied:
Let's defer: I'll create an issue for this. Important right now is testing (which I'm doing on my little frankenstein repo @ https://github.com/Ocramius/just-trying-automated-releases ) and then I think we should just ship this :)

Ocramius commented on src/Github/CreateReleaseTextViaKeepAChangelog.php:
```diff
@@ -0,0 +1,60 @@
+        BranchName $sourceBranch,
+        string $repositoryDirectory
+    ): string {
+        /** @psalm-var non-empty-string $changelog */
```
This is not guaranteed by the `ChangelogParser`: let's instead use `Assert::notEmpty()` **after** `$changelog` has been assigned a `string`

weierophinney replied:
:+1:

That said, `ChangelogParser` will raise an exception if no changelog is found and/or it is empty (it's looking for non-empty contents, IIRC). But the assertion makes it bullet-proof.

Ocramius commented on test/unit/Changelog/ReleaseChangelogViaKeepAChangelogTest.php:
```diff
@@ -0,0 +1,212 @@
+        return $repo;
+    }
+
+    private const INVALID_CHANGELOG = <<< 'END'
```
What happens if `CHANGELOG.md` is empty?

weierophinney replied:
Parser raises an exception.

Ocramius commented on test/unit/Git/CommitFileViaConsoleTest.php:
```diff
@@ -0,0 +1,71 @@
+            ->mustRun();
+
+        (new Process(
+            ['git', 'symbolic-ref', 'HEAD', 'refs/heads/1.0.x'],
```
Fancy trick! This is basically a `git switch`, but to a new branch?

weierophinney replied:
Yep! Discovered it when I was trying to figure out how to test this yesterday. What it does is change the symbolic name of the HEAD to what you provide; it's a way to rename the current branch, essentially.

Ocramius reviewed (approved):
Giving my early LGTM here.

The remaining comments are:

 * [x] naming (ZF habits are hard to let go of :D )
 * [x] testing (in CI, the `GIT_*` environment variables are not defined, so all mock/stub repos need to be configured with user/email before committing)
 * [x] some minor adjustments around assertions
 * [x] did you try it out manually somewhere?

weierophinney committed:
refactor: use lcobucci/clock instead of `date()`

Adds lcobucci/clock as a dependency, and updates
`ReleaseChangelogViaKeepAChangelog` to accept a `Clock` instance as a
dependency. This value is then used to obtain the release date.

In the console tool itself, a `SystemClock` is provided; for testing, we
can use a `FrozenClock` instance.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: adds initial CHANGELOG.md file.

Initializes the CHANGELOG.md file, and adds an entry for #18.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: rename `AppendingCreateReleaseTextAggregate` to `ConcatenateMultipleReleaseTexts`

More descriptive, less "patterny" name.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: incorporate feedback for ConcatenateMultipleReleaseTexts

- Renames "strategies" to "releaseTextGenerators" everywhere, and
  "strategy" to "generator".
- Updates typehint for `$releaseTextGenerators` property to
  `@psalm-param non-empty-list<CreateReleaseText>`
- Instead of declaring `$changelog` a non-empty-string, asserts it.
- Adds `static` declaration to arrow function to fix CS issues.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: rename `ReleaseChangelog` to `CommitReleaseChangelog`

Since "release" is used primarily as a noun in this package, the naming
was ambiguous because it _does_ something; as such, added the verb
`Commit` to the naming of the interface and implementations.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: assign text `CHANGELOG.md` to a class constant

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: more commit text when committing changelog

In addition to the subject `{version} readiness`, the commit text now
also includes the text "Updates the CHANGELOG.md to set the release
date."

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: assertions > type declaration

Use `Assert::notEmpty()` instead of `@psalm-var non-empty-string` to
ensure we get back a non-empty changelog.

Additionally, added the same check in `canCreateReleaseText()` so that
it returns false if the release text returned is empty.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on .github/workflows/psalm.yml:
```diff
@@ -57,3 +57,8 @@ jobs:
       - name: "psalm"
         run: "vendor/bin/psalm --shepherd --stats"
+        env:
+            GIT_AUTHOR_NAME: "Laminas Automatic Releases"
```
Eh, no, let's initialize repositories in tests, rather than requiring environment to be set

weierophinney replied:
The error was from running `git init` within a unit test setup, due to missing author/committer in the env...

Ocramius replied:
Aware: the other tests configure a user/email 👍

weierophinney committed:
qa: refactor how `CommitFileViaConsoleTest` initializes dummy repo

Follows same workflow as `PushViaConsole` tests.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius reviewed (approved):
LGTM :ship:
